Internal Load Balancer

carybakker opened the pull request:
Soundtrack of this PR: [link to song that really fits the mood of this PR]()

### Motivation

load balancer task
### In this PR
* adds url load balancing support

[iOS Implement Internal Load Balancer](https://app.asana.com/0/1201602719155430/1200761120709350)

### Future Work
* Additional tests for load balancer logic
* Adding protocol abstraction for load balancer impl

carybakker committed:
Squashed load balancer changes for rebasing

carybakker committed:
Add logic to make sure nextUrl is always different than currentUrl

the-real-adammork commented on Sources/Network/HTTPS/HttpProtocolConnectionFactory.swift:
```diff
@@ -19,7 +19,7 @@ class HttpProtocolConnectionFactory: ProtocolConnectionFactory {
         ConsensusHttpConnection(
                         config: config,
-                        requester: RestApiRequester(requester: requester, baseUrl: config.url),
+                        requester: RestApiRequester(requester: requester, baseUrl: config.urlLoadBalancer.nextUrl()),
```
i think this is a problem because you rotate to "nextUrl" to set baseUrl, but then the config gets passed down further the chain, and will have nextUrl called again in the Inner.init(...), possibly leading to a mismatch

the-real-adammork commented on Sources/Network/GRPC/GrpcChannelManager.swift:
```diff
@@ -13,7 +13,7 @@ final class GrpcChannelManager {
     func channel(for config: ConnectionConfigProtocol) -> GRPCChannel {
         let wrappedTrustRoots = config.trustRoots[.grpc] as? WrappedNIOSSLCertificate
-        return channel(for: config.url, trustRoots: wrappedTrustRoots?.trustRoots)
+        return channel(for: config.currentUrl, trustRoots: wrappedTrustRoots?.trustRoots)
```
thinking this could be another place where a mismatch could occur

the-real-adammork commented on Sources/Network/NetworkConfig.swift:
```diff
@@ -5,20 +5,23 @@
+    private let consensusUrlLoadBalancer: RandomUrlLoadBalancer<ConsensusUrl>
+    private let fogUrlLoadBalancer: RandomUrlLoadBalancer<FogUrl>
+    
+    var fogUrlsDescription: String {
```
don't think these should be member variables on this struct if just for debugging.

the-real-adammork replied:
can we move these string constructions to in-line variables at the call-site ?

carybakker replied:
sure - good idea, will update

carybakker replied:
the URLs are local to the load balancers, which are private members of the NetworkConfig - so the URLs are not accessible...  Will ping you in Slack to discuss how to expose the info for logging, if we want to.

carybakker replied:
Updated NetworkConfig to expose fog and consensus urls (to not expose the load balancers themselves), where the MobileCoinClient extension builds the descriptive string for logging...

the-real-adammork commented:
Im having concern that were managing the rotation to far into the code-path for "creating" a new connection, and this could cause mismatch because we have to manage multiple downstream usage points.

One of my original ideas was to put the rotation in `HttpProtocolConnectionFactory`/`GrpcProtocolConnectionFactory` above the init for each particular connection. We know 100% this code path only gets called once each time a new connection is created. 

Also, we could make this change lower impact by passing the RandomLoadBalancer in as a separate parameter from the config object. So the config remains mostly the same as it is now and we don't have to propagate all the changes in at each of its usage sites.

Ill work on a prototype implementation for my idea

the-real-adammork commented:
suggested changes here - https://github.com/mobilecoinofficial/MobileCoin-Swift/pull/115

the-real-adammork committed:
Move rotation of URL upstream in the init path (#115)

* Move rotation of URL upstream in the init path. Has the benefit of preserving immutable data structures in more code, and isolates the rotation to the point at which a new connection is created for a given service. Seems like a lower impact change

* fix unit tests

carybakker committed:
Change from computed vars to functions in NetworkConfig for specific config creation

carybakker committed:
Moved creation of the config to be back within the factory block

the-real-adammork commented on Sources/Common/Errors.swift:
```diff
@@ -6,6 +6,20 @@
 
 import Foundation
 
+public struct UnexpectedError: Error {
```
Id prefer not to add a new generic error type, also not seeing it used anywhere so maybe its already been replaced ?

the-real-adammork replied:
remove this error ?

the-real-adammork commented on Sources/Network/Connection/Connections/BlockchainConnection.swift:
```diff
@@ -11,37 +11,38 @@ final class BlockchainConnection:
+        
         super.init(
             connectionOptionWrapperFactory: { transportProtocolOption in
+                let rotatedConfig = config.blockchainConfig()
```
agree that function syntax is more readable and correct for whats happening

the-real-adammork reviewed (commented):
Two comments left in-line, then I think we should do some real-world testing to make sure the rotation works. Im ok having this submitted as screenshots from proxyman or similar. Then I think its good to go.

carybakker committed:
Better exposure of info for logging

the-real-adammork commented on Sources/MobileCoinClient.swift:
```diff
@@ -237,18 +237,26 @@ public final class MobileCoinClient {
         return """
-            Consensus url: \(config.networkConfig.consensusUrl.url)
-            Fog url: \(config.networkConfig.fogUrl.url)
+            Consensus urls: \(consensusUrlsDescription)
```
```suggestion
            Consensus urls: \(config.networkConfig.consensusUrls)
```

the-real-adammork commented on Sources/MobileCoinClient.swift:
```diff
@@ -237,18 +237,26 @@ public final class MobileCoinClient {
-            Consensus url: \(config.networkConfig.consensusUrl.url)
-            Fog url: \(config.networkConfig.fogUrl.url)
+            Consensus urls: \(consensusUrlsDescription)
+            Fog urls: \(fogUrlsDescription)
```
```suggestion
            Fog urls: \(config.networkConfig.fogUrls)
```

carybakker committed:
Update Sources/MobileCoinClient.swift

Co-authored-by: Adam Mork <the-real-adammork@users.noreply.github.com>

carybakker committed:
Update Sources/MobileCoinClient.swift

Co-authored-by: Adam Mork <the-real-adammork@users.noreply.github.com>

carybakker committed:
Removed new error that's no longer used after refactoring code

carybakker committed:
Update url rotation to rotate on error

 - integration test is still WIP and is currently just in place to
   'drive' the balance request and verify rotation if initial URL
   happens to be invalid

the-real-adammork commented on Sources/MobileCoinClient.swift:
```diff
@@ -237,18 +237,26 @@ public final class MobileCoinClient {
 extension MobileCoinClient {
     private static func configDescription(accountKey: AccountKeyWithFog, config: Config) -> String {
         let fogInfo = accountKey.fogInfo
+        var fogUrlsDescription: String {
```
thought we weren't doing this here ?

the-real-adammork commented on Sources/Network/Connection/Connections/ConsensusConnection.swift:
```diff
@@ -32,35 +32,54 @@ final class ConsensusConnection:
             targetQueue: targetQueue)
     }
+    
+    func completionResultIntercept(
```
lets factor this out into one place.

you could make this a function on Connection.swift that returns a closure with a signature same as this function, then at the call-site you dont need the closure syntax

the-real-adammork commented on Tests/Integration/IntegrationTestFixtures.swift:
```diff
@@ -84,6 +84,41 @@ extension IntegrationTestFixtures {
+            consensusUrls: [network.consensusUrl],
+            consensusAttestation: network.consensusAttestation(),
+            fogUrls: [network.fogUrl, "fog://invalid.mobilecoin.com"],
+//            fogUrls: ["fog://invalid.mobilecoin.com"],
```
move this string into one of the fixtures files as a constant

carybakker replied:
yup - the tests were WIP, will do... tx

the-real-adammork committed:
Implement closure returning function for completion interception to clean-up callsite, and factor out re-usable code into one place. Uses generics to ensure type consistency

the-real-adammork commented on Sources/Network/Connection/Connection.swift:
```diff
@@ -34,6 +42,22 @@ class Connection<GrpcService: ConnectionProtocol, HttpService: ConnectionProtoco
+            case .success:
+                completion(result)
+            case .failure:
+                logger.debug("ConsensusConnection - rotating config on error")
```
fyi, you don't need to put the filename in the logging statement, its included as a default parameter this is from the implementation:

```swift
    func debug(
        _ message: @autoclosure () -> String,
        metadata: @autoclosure () -> Logging.Logger.Metadata? = nil,
        source: @autoclosure () -> String? = nil,
        sensitive: Bool = false,
        logFunction: Bool = true,
        file: String = #file,
        function: String = #function,
        line: UInt = #line
    ) {
        debug(
            Message(stringLiteral: message()),
            metadata: metadata(),
            source: source(),
            sensitive: sensitive,
            logFunction: logFunction,
            file: file,
            function: function,
            line: line)
    }
```

`#file` `#function` etc.. are special tokens that capture this info.

carybakker replied:
ya, noticed that when reviewing the logs and meant to remove - tx for reminder.

carybakker committed:
Updates as per PR review comments

carybakker committed:
Minor updates

carybakker committed:
Correct grpc test to use grpc instead of http

carybakker committed:
URL load balancing updates for integration testing:

 * create base class (UrlLoadBalancer) to abstract functionality
 * create SequentialUrlLoadBalancer in integration tests for validation
 * created integration test with sequential load balancer and FogView
   query call for review
 * note: currently the GRPC call does not seem to be hitting the
   completion block when configured with the invalid fog url

carybakker committed:
Add back consecutive call
unit test for RandomUrlLoadBalancer

the-real-adammork committed:
WIP use whenComplete instead of whenSuccess. TODO - Take a look at these config parameters: https://github.com/grpc/grpc-swift/blob/main/docs/keepalive.md we can use them by re-configuring the `ClientConnection` object in `GrpcChannelManager.swift` how you get the parameters in there is upto you dependency injection would be best. Also maybe we need more sensible defaults ? Looks like we can define how long to wait with no connections.

carybakker committed:
Added timeout to call options for grpc

carybakker committed:
Added grpc timeout value to all grpc calls (we need to define the actual timeout we want and put it in a common spot for all the grpc connection classes - Attested, Arbitrary, and regular

carybakker committed:
Add url rotation integration tests for all services

carybakker commented:
Closing this PR to replace with new PR after squashing to rebase.
